torch/csrc/api/include/torch/data/ctf/reader.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <torch/data/ctf/utils.h>
+
+#include <cstdio>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace ctf {
+
+// TODO: Should we use Memory mapped files to speed buffering?
+
+/// A sequential text reader to feed CTF parser
+///
+/// C File API was used due to performance constraints
+/// Current implementation caches chunks of data from file in memory
+/// and parses CTF from it. When it gets empty, buffer is refilled and the cycle
+/// is repeated until EOF is reached
+///
+class Reader {
+ public:
+  virtual ~Reader();
+  explicit Reader(const std::string& filename);
+
+  inline bool can_read(void) const {
+    return (!is_buffer_empty() || can_buffer());
+  }
+  inline const char& peek_char(void) {
+    if (is_buffer_empty()) {
+      refill();
+    }
+    if (rewinded_char_) {
+      return previous_char_;
+    } else {
+      return buffer_[buffer_pos_];
+    }
+  }
+  inline const char& get_char(void) {
+    if (buffer_pos_ > 0) {
+      previous_char_ = buffer_[buffer_pos_ - 1];
```

Comment:
It seems like previous_char_ is always calculated but not always in use. This may incur a slightly perf hit if a lot of get_char() is called. How do you feel like only calculate it when we need it (i.e. when rewinded_char_ is true)?